Preserve types during empty container assignment

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #58911

Differential Revision: [D30785623](https://our.internmc.facebook.com/intern/diff/D30785623)

Commit:
Preserve types during empty container assignment

[ghstack-poisoned]

Commit:
Update on "Preserve types during empty container assignment"




[ghstack-poisoned]

Commit:
Update on "Preserve types during empty container assignment"




[ghstack-poisoned]

Commit:
Update on "Preserve types during empty container assignment"




[ghstack-poisoned]

Commit:
Update on "Preserve types during empty container assignment"




[ghstack-poisoned]

eellison commented:
deferring to @gmagogsfm unless there is a specific thing u want comment on

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -1127,15 +1127,6 @@ void AliasDb::makePointerTo(const Value* from, const Value* to) {
   // immutable. `Any` is mutable but can point to an immutable type
   // through refinement
   if (isMutableTypeInternal(from) != isMutableTypeInternal(to)) {
```
didn't look at any of the other code in here, but we should still be making a pointer if there is any type overlap between the sets. please tag me for changes related to alias analysis. this should actually have been changed as part of the Union changes... maybe you could factor out the code from refinement into a "has type overlap" function  or something similar ? it would be nice if we could reuse that logic to peephole optimize away `prim::isinstance`

eellison replied:
`(isMutableTypeInternal(from) != isMutableTypeInternal(to))` we're checking equality not overlap here

ansley replied:
Sorry for the wording choice, I did mean equality

eellison replied:
i think we want to be checking overlap, e.g. Optional[List[int]] should have a pointer to List[int]

Commit:
Update on "Preserve types during empty container assignment"




[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3804,11 +3887,13 @@ struct to_ir {
-          types, nowhere, /*default_to_union=*/true, element_type_hint);
+          types, nowhere, /*default_to_union=*/true, known_elem_type);
 
       if (!type_hint && *unified == AnyType::get()) {
```
In theory, this should never trigger when `default_to_union=True`.

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3804,11 +3887,13 @@ struct to_ir {
+      // We don't want to use `elem_type` as the final argument to
+      // `unifyTypeList` because there's a chance that `elem_type` is
+      // the Tensor default
+      auto known_elem_type =
```
```suggestion
      auto elem_type_hint =
```

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3789,13 +3837,48 @@ struct to_ir {
     // `Any` as a catch-all supertype). Assume `[]` is `List[Tensor]`
     TypePtr elem_type = TensorType::get();
```
```suggestion
    TypePtr inferred_elem_type = TensorType::get();
```

gmagogsfm commented on test/jit/test_union.py:
```diff
@@ -655,3 +655,92 @@ def fn(x: int) -> str:
+                x["bar"] = torch.tensor(3)
+            return x
+
+        self.checkScript(fn, ())
```
Nit:
Use FileCheck on generated IR to make sure right nodes/types are inserted into graph

Commit:
Update on "Preserve types during empty container assignment"




[ghstack-poisoned]

Commit:
Update on "Preserve types during empty container assignment"




[ghstack-poisoned]

Commit:
Update on "Preserve types during empty container assignment"




[ghstack-poisoned]

Commit:
Update on "Preserve types during empty container assignment"




[ghstack-poisoned]

Commit:
Update on "Preserve types during empty container assignment"


Differential Revision: [D30607282](https://our.internmc.facebook.com/intern/diff/D30607282)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30651066).

Commit:
Update on "Preserve types during empty container assignment"


Differential Revision: [D30651066](https://our.internmc.facebook.com/intern/diff/D30651066)

[ghstack-poisoned]

Commit:
Update on "Preserve types during empty container assignment"


Differential Revision: [D30651066](https://our.internmc.facebook.com/intern/diff/D30651066)

[ghstack-poisoned]

Commit:
Update on "Preserve types during empty container assignment"


Differential Revision: [D30651066](https://our.internmc.facebook.com/intern/diff/D30651066)

[ghstack-poisoned]

Commit:
Update on "Preserve types during empty container assignment"


Differential Revision: [D30651066](https://our.internmc.facebook.com/intern/diff/D30651066)

[ghstack-poisoned]

Commit:
Update on "Preserve types during empty container assignment"


Differential Revision: [D30651066](https://our.internmc.facebook.com/intern/diff/D30651066)

[ghstack-poisoned]

Commit:
Update on "Preserve types during empty container assignment"


Differential Revision: [D30651066](https://our.internmc.facebook.com/intern/diff/D30651066)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30785623).

Commit:
Update on "Preserve types during empty container assignment"


Differential Revision: [D30785623](https://our.internmc.facebook.com/intern/diff/D30785623)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30785623).

Commit:
Update on "Preserve types during empty container assignment"


Differential Revision: [D30785623](https://our.internmc.facebook.com/intern/diff/D30785623)

[ghstack-poisoned]

Commit:
Update on "Preserve types during empty container assignment"


Differential Revision: [D30785623](https://our.internmc.facebook.com/intern/diff/D30785623)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30785623).

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30785623).

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30785623).

Commit:
Update on "Preserve types during empty container assignment"


Differential Revision: [D30785623](https://our.internmc.facebook.com/intern/diff/D30785623)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30785623).

Commit:
Update on "Preserve types during empty container assignment"


Differential Revision: [D30785623](https://our.internmc.facebook.com/intern/diff/D30785623)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30785623).

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30785623).

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30785623).

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30785623).

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30785623).
